lib/services/task_list_service.dart

```diff
@@ -23,141 +27,340 @@ class TaskListService with LogMixin, ChangeCallbackProvider {
+      _decodeTaskListCollection(await _readAndDecodeTaskListCollectionCrdt());
+
+  /// Returns false if the task list does not exist
+  Future<bool> upsertTask(String taskListId, Task task) async {
+    l.info('Upsert task ${task.toJson()}');
+    final crdt = await _readAndDecodeTaskListCollectionCrdt();
+    final taskListCrdt = crdt.get(taskListId);
+    if (taskListCrdt == null) return false;
+
+    task.id ??= Uuid().v4();
+    final taskNode = taskListCrdt.getRecord(task.id!);
+    if (taskNode == null || taskNode.isDeleted) {
+      taskListCrdt.put(task.id!, _encodeTask(task, parent: crdt));
+    } else {
+      taskListCrdt.updateValue(
+        task.id!,
+        (currentValue) => _encodeTask(task, parent: crdt, base: currentValue)
+          ..merge(currentValue),
+      );
     }
 
-    return sorted;
+    await _store(crdt);
+    await _syncService.run();
+
+    return true;
   }
 
-  Future upsert(Task task) async {
-    l.info('Upsert task ${task.toJson()}');
-    final id = Uuid().v4();
-    final update = (await _taskListCrdt)
-      ..put(task.id ?? id, task..id = (task.id ?? id));
-    await _store(update);
+  /// Returns the id of the  new task list.
+  ///
+  /// If [ignoreTasks] is set to true, tasks that are contained in this list will not be stored.
+  Future<void> upsertTaskList(
+    TaskList taskList, {
+    bool ignoreTasks = false,
+  }) async {
+    l.info('Upsert task list ${taskList.toJson()}');
+    final crdt = await _readAndDecodeTaskListCollectionCrdt();
+    taskList.id ??= Uuid().v4();
+    final taskListNode = crdt.getRecord(taskList.id!);
+    if (taskListNode == null || taskListNode.isDeleted) {
+      crdt.put(
+        taskList.id!,
+        _encodeTaskList(
+          taskList,
+          parent: crdt,
+          ignoreTasks: ignoreTasks,
+        ),
+      );
+    } else {
+      crdt.updateValue(
+        taskList.id!,
+        (currentValue) => _encodeTaskList(
+          taskList,
+          parent: crdt,
+          base: currentValue,
+          ignoreTasks: ignoreTasks,
+        )..merge(currentValue),
+      );
+    }
+
+    await _store(crdt);
     await _syncService.run();
   }
 
-  Future remove(Task task) async {
-    final update = (await _taskListCrdt)..delete(task.id!);
+  Future<void> removeTaskList(String taskListId) async {
+    final update = (await _readAndDecodeTaskListCollectionCrdt())
+      ..delete(taskListId);
     await _store(update);
     await _syncService.run();
   }
 
-  Future removeByListID(String taskListID) async {
-    List allTasks = await tasks;
-    for (var i = 0; i < allTasks.length; i++) {
-      if (allTasks[i].taskListID == taskListID) {
-        await remove(allTasks[i]);
-      }
-    }
+  Future<void> removeTask(String taskListId, String taskId) async {
+    final update = (await _readAndDecodeTaskListCollectionCrdt())
+      ..get(taskListId)?.delete(taskId);
+    await _store(update);
+    await _syncService.run();
```

Comment:
Does the syncService then decide whether to actually sync or not (e.g. based on the auto-sync configuration)?